File: Sources/UtilityG.m
#import "UtilityG.h"
#import "NSDictionary-Extension.h"
#import "TeX2img-Swift.h"

void runOkPanel(NSString *title, NSString *message, ...)
{
    va_list arguments;
    va_start(arguments, message);
    NSString *msg = [[NSString alloc] initWithFormat:message arguments:arguments];
    NSAlert *alert = [NSAlert new];
    alert.messageText = title;
    alert.informativeText = msg;
    alert.alertStyle = NSAlertStyleInformational;
    [alert addButtonWithTitle:@"OK"];
    [alert runModal];
    va_end(arguments);
}

void runErrorPanel(NSString *message, ...)
{
    va_list arguments;
    va_start(arguments, message);
    NSString *msg = [[NSString alloc] initWithFormat:message arguments:arguments];
    NSAlert *alert = [NSAlert new];
    alert.messageText = localizedString(@"Error");
    alert.informativeText = msg;
    alert.alertStyle = NSAlertStyleCritical;
    [alert addButtonWithTitle:@"OK"];
    [alert runModal];
    va_end(arguments);
}

void runWarningPanel(NSString *message, ...)
{
    va_list arguments;
    va_start(arguments, message);
    NSString *msg = [[NSString alloc] initWithFormat:message arguments:arguments];
    NSAlert *alert = [NSAlert new];
    alert.messageText = localizedString(@"Warning");
    alert.informativeText = msg;
    alert.alertStyle = NSAlertStyleWarning;
    [alert addButtonWithTitle:@"OK"];
    [alert runModal];
    va_end(arguments);
}

BOOL runConfirmPanel(NSString *message, ...)
{
    va_list arguments;
    va_start(arguments, message);
    NSString *msg = [[NSString alloc] initWithFormat:message arguments:arguments];
    NSAlert *alert = [NSAlert new];
    alert.messageText = localizedString(@"Confirm");
    alert.informativeText = msg;
    alert.alertStyle = NSAlertStyleInformational;
    [alert addButtonWithTitle:@"OK"];
    [alert addButtonWithTitle:localizedString(@"Cancel")];
    NSModalResponse result = [alert runModal];
    va_end(arguments);
    
    return (result == NSAlertFirstButtonReturn);
}


BOOL isJapaneseLanguage()
{
    static BOOL isJapanese;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        NSString *locale = NSLocale.preferredLanguages[0];
        isJapanese = [locale isEqualToString:@"ja"] || [locale isEqualToString:@"ja-JP"];
    });
    return isJapanese;
}

NSColor *readColorFromProfile(Profile *profile,
                              NSString *lightModeKey,
                              NSString *darkModeKey,
                              NSColor *defaultColor)
{
    NSColor *theColor = [profile colorForKey:NSApp.isDarkMode ? darkModeKey : lightModeKey];
    if (!theColor) {
        theColor = defaultColor;
    }
    return theColor;
}

NSColor *foregroundColorInProfile(Profile *profile)
{
    return readColorFromProfile(profile,
                                ForegroundColorForLightModeKey,
                                ForegroundColorForDarkModeKey,
                                NSColor.defaultForegroundColor);
}

NSColor *backgroundColorInProfile(Profile *profile)
{
    return readColorFromProfile(profile,
                                BackgroundColorForLightModeKey,
                                BackgroundColorForDarkModeKey,
                                NSColor.defaultBackgroundColor);
}

NSColor *cursorColorInProfile(Profile *profile)
{
    return readColorFromProfile(profile,
                                CursorColorForLightModeKey,
                                CursorColorForDarkModeKey,
                                NSColor.defaultCursorColor);
}

NSColor *braceColorInProfile(Profile *profile)
{
    return readColorFromProfile(profile,
                                BraceColorForLightModeKey,
                                BraceColorForDarkModeKey,
                                NSColor.defaultBraceColor);
}

NSColor *commentColorInProfile(Profile *profile)
{
    return readColorFromProfile(profile,
                                CommentColorForLightModeKey,
                                CommentColorForDarkModeKey,
                                NSColor.defaultCommentColor);
}

NSColor *commandColorInProfile(Profile *profile)
{
    return readColorFromProfile(profile,
                                CommandColorForLightModeKey,
                                CommandColorForDarkModeKey,
                                NSColor.defaultCommandColor);
}

NSColor *invisibleColorInProfile(Profile *profile)
{
    return readColorFromProfile(profile,
                                InvisibleColorForLightModeKey,
                                InvisibleColorForDarkModeKey,
                                NSColor.defaultInvisibleColor);
}

NSColor *highlightedBraceColorInProfile(Profile *profile)
{
    return readColorFromProfile(profile,
                                HighlightedBraceColorForLightModeKey,
                                HighlightedBraceColorForDarkModeKey,
                                NSColor.defaultHighlightedBraceColor);
}

NSColor *enclosedContentBackgroundColorInProfile(Profile *profile)
{
    return readColorFromProfile(profile,
                                EnclosedContentBackgroundColorForLightModeKey,
                                EnclosedContentBackgroundColorForDarkModeKey,
                                NSColor.defaultEnclosedContentBackgroundColor);
}

NSColor *flashingBackgroundColorInProfile(Profile *profile)
{
    return readColorFromProfile(profile,
                                FlashingBackgroundColorForLightModeKey,
                                FlashingBackgroundColorForDarkModeKey,
                                NSColor.defaultFlashingBackgroundColor);
}

NSColor *consoleForegroundColorInProfile(Profile *profile)
{
    return readColorFromProfile(profile,
                                ConsoleForegroundColorForLightModeKey,
                                ConsoleForegroundColorForDarkModeKey,
                                NSColor.defaultConsoleForegroundColor);
}

NSColor *consoleBackgroundColorInProfile(Profile *profile)
{
    return readColorFromProfile(profile,
                                ConsoleBackgroundColorForLightModeKey,
                                ConsoleBackgroundColorForDarkModeKey,
                                NSColor.defaultConsoleBackgroundColor);
}

File: Sources/Utility/UtilityG.swift
import AppKit
import Foundation

@objc(UtilityG)
class UtilityG: NSObject {
    @objc(runOkPanelWithTitle:message:)
    static func runOkPanel(title: String, message: String) {
        let alert = NSAlert()
        alert.messageText = title
        alert.informativeText = message
        alert.alertStyle = .informational
        alert.addButton(withTitle: "OK")
        alert.runModal()
    }

    @objc(runErrorPanelWithMessage:)
    static func runErrorPanel(message: String) {
        let alert = NSAlert()
        alert.messageText = NSLocalizedString("Error", comment: "")
        alert.informativeText = message
        alert.alertStyle = .critical
        alert.addButton(withTitle: "OK")
        alert.runModal()
    }

    @objc(runWarningPanelWithMessage:)
    static func runWarningPanel(message: String) {
        let alert = NSAlert()
        alert.messageText = NSLocalizedString("Warning", comment: "")
        alert.informativeText = message
        alert.alertStyle = .warning
        alert.addButton(withTitle: "OK")
        alert.runModal()
    }

    @objc(runConfirmPanelWithMessage:)
    static func runConfirmPanel(message: String) -> Bool {
        let alert = NSAlert()
        alert.messageText = NSLocalizedString("Confirm", comment: "")
        alert.informativeText = message
        alert.alertStyle = .informational
        alert.addButton(withTitle: "OK")
        alert.addButton(withTitle: NSLocalizedString("Cancel", comment: ""))
        return alert.runModal() == .alertFirstButtonReturn
    }

    @objc static func isJapaneseLanguage() -> Bool {
        struct Cache {
            static let value: Bool = {
                guard let locale = Locale.preferredLanguages.first else { return false }
                return locale == "ja" || locale == "ja-JP"
            }()
        }
        return Cache.value
    }

    private static func readColor(from profile: NSDictionary,
                                  lightModeKey: String,
                                  darkModeKey: String,
                                  defaultColor: NSColor) -> NSColor {
        let key = NSApp.isDarkMode ? darkModeKey : lightModeKey
        return profile.colorForKey(key) ?? defaultColor
    }

    @objc(foregroundColorInProfile:)
    static func foregroundColor(inProfile profile: NSDictionary) -> NSColor {
        readColor(from: profile,
                  lightModeKey: "foregroundColor",
                  darkModeKey: "foregroundColorForDarkMode",
                  defaultColor: .defaultForegroundColor)
    }

    @objc(backgroundColorInProfile:)
    static func backgroundColor(inProfile profile: NSDictionary) -> NSColor {
        readColor(from: profile,
                  lightModeKey: "backgroundColor",
                  darkModeKey: "backgroundColorForDarkMode",
                  defaultColor: .defaultBackgroundColor)
    }

    @objc(cursorColorInProfile:)
    static func cursorColor(inProfile profile: NSDictionary) -> NSColor {
        readColor(from: profile,
                  lightModeKey: "cursorColor",
                  darkModeKey: "cursorColorForDarkMode",
                  defaultColor: .defaultCursorColor)
    }

    @objc(braceColorInProfile:)
    static func braceColor(inProfile profile: NSDictionary) -> NSColor {
        readColor(from: profile,
                  lightModeKey: "braceColor",
                  darkModeKey: "braceColorForDarkMode",
                  defaultColor: .defaultBraceColor)
    }

    @objc(commentColorInProfile:)
    static func commentColor(inProfile profile: NSDictionary) -> NSColor {
        readColor(from: profile,
                  lightModeKey: "commentColor",
                  darkModeKey: "commentColorForDarkMode",
                  defaultColor: .defaultCommentColor)
    }

    @objc(commandColorInProfile:)
    static func commandColor(inProfile profile: NSDictionary) -> NSColor {
        readColor(from: profile,
                  lightModeKey: "commandColor",
                  darkModeKey: "commandColorForDarkMode",
                  defaultColor: .defaultCommandColor)
    }

    @objc(invisibleColorInProfile:)
    static func invisibleColor(inProfile profile: NSDictionary) -> NSColor {
        readColor(from: profile,
                  lightModeKey: "invisibleColor",
                  darkModeKey: "invisibleColorForDarkMode",
                  defaultColor: .defaultInvisibleColor)
    }

    @objc(highlightedBraceColorInProfile:)
    static func highlightedBraceColor(inProfile profile: NSDictionary) -> NSColor {
        readColor(from: profile,
                  lightModeKey: "highlightedBraceColor",
                  darkModeKey: "highlightedBraceColorForDarkMode",
                  defaultColor: .defaultHighlightedBraceColor)
    }

    @objc(enclosedContentBackgroundColorInProfile:)
    static func enclosedContentBackgroundColor(inProfile profile: NSDictionary) -> NSColor {
        readColor(from: profile,
                  lightModeKey: "enclosedContentBackgroundColor",
                  darkModeKey: "enclosedContentBackgroundColorForDarkMode",
                  defaultColor: .defaultEnclosedContentBackgroundColor)
    }

    @objc(flashingBackgroundColorInProfile:)
    static func flashingBackgroundColor(inProfile profile: NSDictionary) -> NSColor {
        readColor(from: profile,
                  lightModeKey: "flashingBackgroundColor",
                  darkModeKey: "flashingBackgroundColorForDarkMode",
                  defaultColor: .defaultFlashingBackgroundColor)
    }

    @objc(consoleForegroundColorInProfile:)
    static func consoleForegroundColor(inProfile profile: NSDictionary) -> NSColor {
        readColor(from: profile,
                  lightModeKey: "consoleForegroundColor",
                  darkModeKey: "consoleForegroundColorForDarkMode",
                  defaultColor: .defaultConsoleForegroundColor)
    }

    @objc(consoleBackgroundColorInProfile:)
    static func consoleBackgroundColor(inProfile profile: NSDictionary) -> NSColor {
        readColor(from: profile,
                  lightModeKey: "consoleBackgroundColor",
                  darkModeKey: "consoleBackgroundColorForDarkMode",
                  defaultColor: .defaultConsoleBackgroundColor)
    }
}